python/lsst/ctrl/mpexec/singleQuantumExecutor.py

```diff
@@ -0,0 +1,161 @@
+#  Imports of standard modules --
+# -------------------------------
+import logging
+
+# -----------------------------
+#  Imports for other modules --
+# -----------------------------
+
+# ----------------------------------
+#  Local non-exported definitions --
+# ----------------------------------
+
+_LOG = logging.getLogger(__name__.partition(".")[2])
+
+
+class SingleQuantumExecutor:
+    """Executor class which runs one Quantum at a time.
+
+    Parameters
+    ----------
+    butler : `~lsst.daf.butler.Butler`
+        Data butler.
+    taskFactory : `~lsst.pipe.base.TaskFactory`
+        Instance of a task factory.
+    """
+    def __init__(self, butler, taskFactory):
+        self.butler = butler
+        self.taskFactory = taskFactory
+
+    def execute(self, taskClass, config, quantum):
+        """Execute super-task on a single Quantum.
+
+        Parameters
+        ----------
+        taskClass : `type`
+            Sub-class of `~lsst.pipe.base.PipelineTask`.
+        config : `~lsst.pipe.base.PipelineTaskConfig`
+            Configuration object for this task
+        quantum : `~lsst.daf.butler.Quantum`
+            Single Quantum instance.
+        """
+        self.setupLogging(taskClass, config, quantum)
+        self.updateQuantumInputs(quantum)
+        task = self.makeTask(taskClass, config)
+        self.runQuantum(task, quantum)
+        self.saveQuantum(quantum, taskClass)
+
+    def setupLogging(self, taskClass, config, quantum):
+        """Configure logging system for execution of this task.
+
+        Ths method can setup logging to attach task- or
+        quantum-specific information to log messages. Potentially this can
+        take into accout some info from task configuration as well.
+
+        Parameters
+        ----------
+        taskClass : `type`
+            Sub-class of `~lsst.pipe.base.PipelineTask`.
+        config : `~lsst.pipe.base.PipelineTaskConfig`
+            Configuration object for this task
+        quantum : `~lsst.daf.butler.Quantum`
+            Single Quantum instance.
+        """
+        # include dataId into MDC
+#         if dataRef is not None:
+#             if hasattr(dataRef, "dataId"):
+#                 lsst.log.MDC("LABEL", str(dataRef.dataId))
+#             elif isinstance(dataRef, (list, tuple)):
+#                 lsst.log.MDC("LABEL", str([ref.dataId for ref in dataRef if hasattr(ref, "dataId")]))
+
+    def makeTask(self, taskClass, config):
+        """Make new task instance.
+
+        Parameters
+        ----------
+        taskClass : `type`
+            Sub-class of `~lsst.pipe.base.PipelineTask`.
+        config : `~lsst.pipe.base.PipelineTaskConfig`
+            Configuration object for this task
+
+        Returns
+        -------
+        task : `~lsst.pipe.base.PipelineTask`
+            Instance of ``taskClass`` type.
+        """
+        # call task factory for that
+        return self.taskFactory.makeTask(taskClass, config, None, self.butler)
```

Comment:
Not sure what do you mean by custom logging and what info do you want to pass to a task constructor, do you have some examples?

We can certainly factor out implementation of the `makeTask` from `TaskFactory` - it does not need that dependency on a TaskLoader so it may be beneficial to be able to use that code without instantiating TaskLoader.

Reply:
I was going to say that `Task.__init__` takes a `lsst.log.Log` to be used for all of the logging it and its subtasks do, so if you want to customize that you'd need to do it here.

But looking closer, it seems it just uses that to set the `Log` instance's name (which we probably don't want to change here), and I presume that means customization (e.g. making all log messages include the data ID) would happen via some other API that doesn't work directly on the `Log` instance.